File: game/src/game/Game.ts

type PlayRecord = {
    [key in CellValue]: number[][] | null;
}

export class Game {
    private static readonly INITIAL_PLAYER: CellValue = CellValue.RED;

    private board: GameBoard;
    private currentPlayer: CellValue = Game.INITIAL_PLAYER;

    private playRecord: PlayRecord = {
        [CellValue.RED]: [],
        [CellValue.BLUE]: [],
        [CellValue.EMPTY]: null
    };

    constructor(size: number) {
        this.board = new GameBoard(size);
    }

    getBoard(): GameBoard {
        return this.board;
    }

    getCurrentPlayer(): CellValue {
        return this.currentPlayer;
    }

    play(row: number, col: number): MoveResult {
        const cell = this.board.getCell(row, col);
        if (cell !== CellValue.EMPTY) {
            return MoveResult.INVALID;
        }

        this.board.setCell(row, col, this.currentPlayer);
        this.playRecord[this.currentPlayer].push([row, col]);

        if (this.checkWinner()) {
            if (this.currentPlayer === CellValue.RED)
                return MoveResult.RED_WIN;
            return MoveResult.BLUE_WIN;
        }

        this.swapPlayer();

        return MoveResult.VALID;
    }

    checkWinner(): boolean {
        if (this.playRecord[this.currentPlayer].length < this.board.getSize())
            return false;

        let visited: boolean[][] = this.fill2DArray(this.board.getSize(), this.board.getSize(), false);
        
        // all nodes of current player
        for (let [row, col] of this.playRecord[this.currentPlayer]) {
            // dfs on boundary nodes only
            let redStartEdge: boolean = this.currentPlayer === CellValue.RED && col === 0;
            let blueStartEdge: boolean = this.currentPlayer === CellValue.BLUE && row === 0;

            if (visited[row][col] || !(redStartEdge || blueStartEdge))
                continue;

            if (this.dfsOnNode(row, col, visited))
                return true;
        }

        return false;
    }

    private dfsOnNode(row: number, col: number, visited: boolean[][]): boolean {
        visited[row][col] = true;
        let currentStack: number[][] = [[row, col]];

        while (currentStack.length > 0) {
            let [r, c] = currentStack.pop()!;

            let neighbours = this.board.getNeighbours(r, c);
            for (let [nr, nc] of neighbours) {
                if (this.board.getCell(nr, nc) !== this.currentPlayer)
                    continue;

                // check if winning node is found
                let redEndEdge: boolean = this.currentPlayer === CellValue.RED
                        && nc === this.board.getSize() - 1;
                let blueEndEdge: boolean = this.currentPlayer === CellValue.BLUE
                        && nr === this.board.getSize() - 1;
                if (redEndEdge || blueEndEdge)
                    return true;

                // continue dfs
                if (!visited[nr][nc]) {
                    visited[nr][nc] = true;
                    currentStack.push([nr, nc]);
                }
            }
        }
    }

    private fill2DArray<T>(rows: number, cols: number, value: T): T[][] {
        return Array.from({ length: rows }, () => Array(cols).fill(value));
    }

    reset(): void {
        this.board.reset();
        this.currentPlayer = Game.INITIAL_PLAYER;

        this.playRecord[CellValue.RED] = [];
        this.playRecord[CellValue.BLUE] = [];
    }

    swapPlayer(): void {
        this.currentPlayer = this.currentPlayer === CellValue.RED ? CellValue.BLUE : CellValue.RED;
    }
}

File: game/src/game/game.rs
use crate::game::cell::CellValue;
use crate::game::board::Board;
use crate::game::move_result::MoveResult;

use std::collections::HashMap;

pub struct Game {
    board: Board,
    current_player: CellValue,
    play_record: HashMap<CellValue, Vec<(usize, usize)>>,
}

impl Game {
    const INITIAL_PLAYER: CellValue = CellValue::Red;

    pub fn new(size: usize) -> Game {
        let board = Board::new(size);
        let current_player = Game::INITIAL_PLAYER;
        let mut play_record = HashMap::new();

        play_record.insert(CellValue::Red, Vec::new());
        play_record.insert(CellValue::Blue, Vec::new());

        Game { board, current_player, play_record }
    }

    pub fn get_board(&self) -> &Board {
        &self.board
    }

    pub fn get_current_player(&self) -> CellValue {
        self.current_player
    }

    pub fn play(&mut self, row: usize, col: usize) -> MoveResult {
        let cell: CellValue = self.board.get_cell(row, col);

        if cell != CellValue::Empty {
            return MoveResult::Invalid;
        }

        self.board.set_cell(row, col, self.current_player);
        self.play_record
            .get_mut(&self.current_player)
            .unwrap()
            .push(
                (row, col)
            );

        if self.check_winner() {
            match self.current_player {
                CellValue::Red => return MoveResult::RedWin,
                CellValue::Blue => return MoveResult::BlueWin,
                _ => return MoveResult::Invalid,
            }
        }

        self.swap_player();

        MoveResult::Valid
    }

    fn check_winner(&self) -> bool {
        if self.play_record.get(&self.current_player).unwrap().len() < self.board.get_size() {
            return false;
        }

        let mut visited: Vec<Vec<bool>> = vec![vec![false; self.board.get_size()]; self.board.get_size()];

        for (row, col) in self.play_record.get(&self.current_player).unwrap() {
            let red_start_edge: bool = self.current_player == CellValue::Red && *col == 0;
            let blue_start_edge: bool = self.current_player == CellValue::Blue && *row == 0;

            if visited[*row][*col] || !(red_start_edge || blue_start_edge) {
                continue;
            }

            if self.dfs_on_node(*row, *col, &mut visited) {
                return true;
            }
        }

        false
    }

    fn dfs_on_node(&self, row: usize, col: usize, visited: &mut Vec<Vec<bool>>) -> bool {
        visited[row][col] = true;
        let mut stack: Vec<(usize, usize)> = Vec::new();

        stack.push((row, col));

        while stack.len() > 0 {
            let (r, c) = stack.pop().unwrap();

            let neighbours = self.board.get_neighbours(r, c);
            for (nr, nc) in neighbours {
                if self.board.get_cell(nr, nc) != self.current_player {
                    continue;
                }

                let end_edge = match self.current_player {
                    CellValue::Red => nc == self.board.get_size() - 1,
                    CellValue::Blue => nr == self.board.get_size() - 1,
                    _ => false,
                };

                if end_edge {
                    return true;
                }

                if !visited[nr][nc] {
                    visited[nr][nc] = true;
                    stack.push((nr, nc));
                }
            }
        }

        false
    }

    pub fn reset(&mut self) -> () {
        self.board.reset();
        self.current_player = Game::INITIAL_PLAYER;

        self.play_record
            .get_mut(&CellValue::Red)
            .unwrap()
            .clear();
        self.play_record
            .get_mut(&CellValue::Blue)
            .unwrap()
            .clear();
    }

    fn swap_player(&mut self) -> () {
        match self.current_player {
            CellValue::Red => self.current_player = CellValue::Blue,
            CellValue::Blue => self.current_player = CellValue::Red,
            _ => (),
        }
    }
}